recipe/meta.yaml

```diff
@@ -56,6 +62,11 @@ build:
   # 2024/04 This build seems to fail due to disk space on the CIs
   # We can potentially re-enable it at a later time
   skip: true  # [cuda_compiler_version == "11.8" and aarch64]
+{% if "rc" in version %}
+  skip: true  # [rc == "None"]
+{% else %}
+  skip: true  # [rc == "rc"]
+{% endif %}
```

Comment:
What's the point of this? We're only building one or the other anyway (depending on what's in `version`), so I'd just delete this (and the `rc:` variable in CBC) completely

Reply:
It is used to select whether and RC label is used for channel upload. Please see below

Reply:
yes. Are you ok with this schema?

Reply:
There's something I'm missing (after reading docs and related issues).

AFAIU:
- the `_numpy_rc` package is given the label `numpy_rc` in the conda-forge channel
- the pattern above means that if an rc build of numpy is desired, the `_numpy_rc` package, labelled `numpy_rc` is used
- due to strict channel priority, if the `numpy_rc` label is included as a higher-priority channel_source, any packages labelled `numpy_rc` are considered, and for those packages, versions/variants that aren't labelled `numpy_rc` are not considered.

I don't understand how this would affect channel_targets of numpy's rc builds though. Does channel_targets automatically change to `conda-forge numpy_rc` if `_numpy_rc` is pulled, somehow? Also, I can't see any numpy packages labelled `numpy_rc` on the [conda-forge channel](https://anaconda.org/conda-forge/numpy/labels), and I can see rc versions being uploaded [without a `numpy_rc` label](https://anaconda.org/conda-forge/numpy/files).

Reply:
I also don't see how deleting the `rc` variable in cbc.yaml would work, if we want to have two channel_targets (which would need to be zipped with something, I guess, otherwise everything would be built twice and uploaded to both labels)

Reply:
I'm still getting familiar with conda-forge's CI/smithy/rendering etc, so thanks for any info!

Reply:
> I don't understand how this would affect channel_targets of numpy's rc builds though. 

That's the neat trick about this. They can be published to `main`, and only the "magic" package `_numpy_rc` needs to be under a different channel  (under the opt-in `conda-forge/label/numpy_rc`)
 
> Does channel_targets automatically change to `conda-forge numpy_rc` if `_numpy_rc` is pulled, somehow? Also, I can't see any numpy packages labelled `numpy_rc` on the [conda-forge channel](https://anaconda.org/conda-forge/numpy/labels), and I can see rc versions being uploaded [without a `numpy_rc` label](https://anaconda.org/conda-forge/numpy/files).

How this works is that:
```
conda install numpy=2.2.0rc1  # fails due to missing dependency _numpy_rc
conda install -c conda-forge/label/numpy_rc numpy=2.2.0rc1  # succeeds
```

This isn't necessary per se for pytorch, but it's a less intrusive approach to handle rc-builds IMO (at the cost of having an extra magic package).

Reply:
> This isn't necessary per se for pytorch, but it's a less intrusive approach to handle rc-builds IMO (at the cost of having an extra magic package).

Let's not do this for all packages. The accepted behaviour is to put the RC package into `conda-forge/label/pytorch_rc` as specified in the CFEP. The magic package behaviour is to work-around some limitations with conda-forge CI infrastrcture when building new python/numpy packages based on the RCs. Since we are not going to build downstreams with one released version and one rc, this is not needed and adds extra complexity.

Reply:
Fine by me!

Reply:
ok, thanks for the explanation, I understand that now. It stops RCs from being installed, unless you specify that label with `-c`.
@h-vetinari @jakirkham are you good with this as-is? Any further comments?

Reply:
Apart from removing the `rc1` in the version string

Reply:
> are you good with this as-is? Any further comments?

The main remaining question to me is whether we're going to be building rc's so often that it's justified to add this infrastructure to `main`. If the answer is yes: fine by me. Historically, this feedstock has moved slowly enough that rc's were built (if at all) in PRs, and could then turn into a regular version bump as soon as the version went GA.

Reply:
I mentioned this at the start of the PR:

> What are your requirements around this in general? We use it to prepare for the main release and reduce latency between upstream and the conda package. We don't upload the rc packages.

Since I'm pretty new to understanding conda-forge culture/requirements, I went with the guidance from @hmaarrfk and @jakirkham . Assuming that what they were asking reflected the intention behind the first bullet point [here](https://github.com/conda-forge/pytorch-cpu-feedstock/issues/273).

Reply:
Personally am good with what we have here

Agree with Isuru that we likely don't need the `pytorch_rc`  package approach here. That would only be useful if we want to migrate to a PyTorch RC in conda-forge and we think there will be enough time between the RC and the final release to merit that effort. Personally am not convinced we need this additional complexity. Could be convinced if we start packaging RCs more regularly and quickly

My understanding is this addition of RC builds was helpful also for alignment with Anaconda's package. So tried to recommend a minimal solution to accomplish that goal

Think it would be good to hear from Mark on these changes

Reply:
I agree. Additional to what we will pull across, this includes a minimal mechanism to stop rc builds being used by normal users too. So seems minimal as-is.
@hmaarrfk any comments?

Reply:
I mostly don't want to give "an other" opinion.

I think that `rc` builds are generally helpful to anticipate build problems.


The additions here seem "minimal" and well thought out.

it also gives us the ability to give feedback to upstream 

But with 2Month+ merge times on 1 line fixes:
* https://github.com/pytorch/pytorch/pull/137331
* https://github.com/pytorch/pytorch/pull/137084

I'm personally not so enthusiastic about finding "suggestions" during the `rc` and having them merged in before the full release.

Ultimately, I'm very glad to see some new contributors to this feedstock. If you think the `rc` builds would empower you to continue to contribute (perhaps by using this feedstock and uploading to your own channel) that is enough for me to accept this in.

We can always "remove" the `rc` stuff later.

Reply:
It's less about providing upstream fixes, and more about getting the feedstock to build ahead of the release date, so that it's available to conda users with minimal latency compared to the pypi package.
Which is especially useful now that pytorch have dropped conda packages.

Reply:
I basically want to merge anaconda's recipe with this one a bit more. And thought this would be a relatively trivial, easy-first-issue sort of change 😆

Reply:
I'm learning that conda-forge's recipes, being public, need to be gone over more than Anaconda's ones, because they'll be used by third parties and need to be more foolproof

Reply:
basically for me you've made it pretty clear:

> danpetry is willing to help. he has taken the initiative to work through concerns of adding an `rc` and shown to be receptive to feedback. I think this is worth a try to help unify the recipe and get him invested in the maintenance.


Regarding:

> I'm learning that conda-forge's recipes, being public, need to be gone over more than Anaconda's ones

Its mostly that this recipe is very popular by end users. and it is performance critical. Many core members have used it in their work (myself included) so its just a little more touchy.

We also require many special resources, CFEP03 or self-hosted runners, which start to necessitate coordination.

Reply:
My intention is that in the future we can pool resources rather than develop this recipe separately